lib/slack/client.ex

```diff
@@ -0,0 +1,257 @@
+defmodule Slack.Client do
+  @moduledoc """
+  A datastructure that allows you to monitor, inspect, and react to the state of
+  your Slack team.
+  """
+  
+  if :lt == Version.compare(System.version, "1.1.0") do
+    
+    @derive [Access]
+    
+  else
+  
+    @behaviour Access
+    
+    @type key :: any
+    @type value :: any
+
+    @doc """
+    Implements fetch/2 for Access protocol.
+    """
+    @spec fetch(Client.t, key) :: value
+    def fetch(client, key)
+    defdelegate fetch(client, key), to: Map
+      
+    @doc """
+    Implements get_and_update/3 for Access protocol.
+    """
+    @spec get_and_update(Client.t, key, (value -> {value, value})) :: Client.t
+    def get_and_update(client, key, function)
+    defdelegate get_and_update(client, key, function), to: Map
+    
+  end
+  
+  defstruct [
+    socket: nil,
+    client: nil,
+    token:  nil,
+    me:     nil,
+    team:   nil,
+    bots:     %{},
+    channels: %{},
+    groups:   %{},
+    users:    %{},
+    ims:      %{},
+  ]
+    
+  @doc """
+  Notice when a new Slack object has been created.
```

Comment:
I don't think `Notice` is correct. `Update client state when an update is received`?